Convert the StackFrameList mutex to a shared mutex.

jimingham opened the pull request:
In fact, there's only one public API in StackFrameList that changes
 the list explicitly.  The rest only change the list if you happen to
ask for more frames than lldb has currently fetched and that 
always adds frames "behind the user's back".  So we were
much more prone to deadlocking than we needed to be.

This patch uses a shared_mutex instead, and when we have to add more
frames (in GetFramesUpTo) we switches to exclusive long enough to add
the frames, then goes back to shared.
    
Most of the work here was actually getting the stack frame list locking to not
require a recursive mutex (shared mutexes aren't recursive). 
    
I also added a test that has 5 threads progressively asking for more
frames simultaneously to make sure we get back valid frames and don't
deadlock.

jimingham committed:
Convert the recursive StackFrameList mutex to a non-recursive one.

jimingham committed:
Convert the StackFrameList mutex to a shared mutex.

In fact, there's only one public API in StackFrameList that changes
the list explicitly.  The rest only change the list if you happen to
ask for more frames than lldb has currently fetched.  So we were
much more prone to deadlocking than we needed to be.

This patch uses a shared_mutex instead, and when we have to add more
frames (in GetFramesUpTo) we switches to exclusive long enough to add
the frames, then goes back to shared.

I also added a test that has 5 threads progressively asking for more
frames simultaneously to make sure we get back valid frames and don't
deadlock.

jimingham committed:
Formatting.

labath commented on lldb/source/Target/StackFrameList.cpp:
```diff
@@ -771,24 +809,38 @@ void StackFrameList::SelectMostRelevantFrame() {
-  std::lock_guard<std::recursive_mutex> guard(m_mutex);
+  uint32_t result = 0;
+  {
+    std::shared_lock<std::shared_mutex> guard(m_list_mutex);
+    result = SetSelectedFrameNoLock(frame);
```
A shared lock around a function called "Set" is very suspicious. What is this trying to protect?

jimingham replied:
The list mutex protects the stack frame list, not metadata like the selected frame pointer.  So this function is a reader of the stack frame list, though it will update the metadata.

labath commented on lldb/source/Target/StackFrameList.cpp:
```diff
@@ -128,30 +130,41 @@ bool StackFrameList::DecrementCurrentInlinedDepth() {
-                                               Unwind &unwinder) {
+void StackFrameList::GetOnlyConcreteFramesUpTo(
+    uint32_t end_idx, Unwind &unwinder,
+    std::shared_lock<std::shared_mutex> &guard) {
```
All of the work of this function is done under the exclusive mutex. Why does it need the lock dance?

jimingham replied:
I don't understand this comment.  In the patch as posted, GetOnly ConcreteFramesUpTo gets called before the shared lock is dropped and the exclusive mutex acquired (since it returned right away in its only use, it seemed nicer to put that dance in the function.  That should be clear by moving the equivalent "get concrete and inlined frames" code into its own function.

labath reviewed (commented):
This switching of lock types is making me very nervous. Every switch is an opportunity to get preempted, which means you have to recheck any invariants before you do that -- you can't just assume that conditions you tested earlier in the function still hold.

That essentially forces you to treat all work being done under one lock type as a single unit -- with well defined inputs, outputs, pre- and post-conditions (okay, maybe it doesn't, but I'd argue that if you can't say what is the outcome of a block of code like this, then its very hard to guarantee or check that its being used correctly).

Now, if we have a piece of code with inputs and outputs and pre- and post-conditions, then the best thing to do is to put that in a function. For something like stack unwinding, my ideal setup would be to have one function for the "shared" work, one for the "exclusive" work and then one to orchestrate everything (each one can be further subdivided as needed).
```
GetFramesUpTo(idx) {
  frame = GetCachedFrame(idx); // uses read/shared lock internally
  if (!frame) {
    // uses write/exclusive lock internally, cannot assume that frame[idx] does not exist as it might have been created in the meantime.
    frame = GetOrCreateFrame(idx); 
  }
  // Do something with the frame (with or without a lock).
}
```

So, my question is, why can't the code be structured to look something like this (I know my example is too simple to work verbatim, but it's just an illustration)? If not, then why, and how is one supposed to reason about code like this (these functions are sufficiently complex without needing to worry about lock types and preemption)? I think it's totally reasonable and desirable (and expected?) to precede a patch like this with some refactor which makes the code more amenable to a different locking strategy.

jimingham committed:
Refactor the second half of GetFramesUpTo (the part that required the
exclusive mutex) into a separate function for clarity.

jimingham commented:
Does this look better?

The situation is that there are really only three "writer" API's: `GetOnlyConcreteFramesUpTo`,` `Clear` and the second half of `GetFramesUpTo`, and only Clear is called with the intention of changing the list contents.  And none of these were client API's.  All the client API's were formally readers - they shouldn't have to know that we update frames lazily.  So the responsibilities were already pretty strictly limited.

But clearer is for sure better, so I renamed `GetOnlyConcreteFramesUpTo` to `FetchOnlyConcreteFramesUpTo` and moved the second half of GetFramesUpTo into `FetchFramesUpTo`.  Those and Clear are the only places where the list is actually mutated.  `Clear` doesn't need to do any dance, since it's a client API and intends to change the list.  Other than that, clients who want to read the list need to acquire the shared mutex, if the `Fetch` functions need to add frames to fulfill that request, that should happen silently when some client asks for more frames than have been fetched.

labath reviewed (commented):
It's somewhat better, but it doesn't address my main concern, which is this mid-scope switching of lock types. It's an unusual setup, and that's usually not good thing when it comes to locking and threads.

I don't think that the fact that these are all reader APIs (for clients of StackFrameList) has anything to do with this, as the locking scheme is an implementation detail of the StackFrameList class (evidenced by the fact that you can change it without modifying any of the clients). The interface we're concerned with here is the one between `m_frames` member (and anything else being protected) and the code accessing it. And there are definitely writers there.

I still believe that for a code like this to be correct, each piece of code covered by a single type of a lock needs to have a clearly defined purpose. And if it has a purpose, then it can be written as a scope (block or a function).

Let's take a look at `FetchOnlyConcreteFramesUpTo` and `FetchFramesUpTo`. Their only caller is `GetFramesUpTo` and their called close together, which means scoping them should be fairly easy.

Now, let's see what happens before this "write" block. Inside GetNumFrames, we basically only have a "have the frames been computed already" check, which is amenable to be put into a separate function block. `GetFramesUpTo` is called from two places. One of them is `GetNumFrames` which does nothing besides locking the mutex and them calling `GetFramesUpTo`. The other is `GetFrameAtIndexNoLock` which does another "has the frame been computed check" (is there any difference between the two, and can they be merged?) and then calls `GetFramesUpTo`. It doesn't lock the mutex though, so lets look at where that happens.

This happens either in `GetFrameAtIndex`, where the pattern is again "lock the mutex and call the next function"; in `GetFrameWithConcreteFrameIndex`, which AFAICT is not doing anything that needs protecting, as its only working with the frames returned from `GetFrameAtIndex`; or in `GetFrameWithStackID`, which is doing a slightly different form of "has this frame been computed" check (via a binary search), and then proceeds to generate new frames (it starts calling `GetFrameAtIndexNoLock` with frame zero, but I think it could start with m_frames.size(), since it has checked the previous frames already).

So it seems to me that at least everything that happens before taking the write lock could be expressed as some sort of a function or a block. I haven't looked that closely at what happens after releasing the write lock, but I believe the same could be done there as well (with the exception of one block, which I think actually needs to be protected by the write lock as well -- see inline comment). I think the code would look better if we did that.

jimingham committed:
Remove code that tried to set concrete frames after GetFramesUpTo.

That code was wrong, it was prefaced by the comment that only when
we get inlined frames does GetFramesUpTo fetch the needed frames, but
that's not true.

jimingham commented on lldb/source/Target/StackFrameList.cpp:
```diff
@@ -594,7 +643,7 @@ StackFrameSP StackFrameList::GetFrameAtIndex(uint32_t idx) {
+  if (GetFramesUpTo(idx, AllowInterruption, guard)) {
     Log *log = GetLog(LLDBLog::Thread);
     LLDB_LOG(log, "GetFrameAtIndex was interrupted");
     return {};
```
You have sharp eyes, thanks!  I missed that code.  It's based on a false premise (GetFramesUpTo fills the stack frames up to the depth requested in both the "concrete frames only" case and the "inlined frames".  So this code is redundant (and will always find the required frames filled by the time you get to it.)
However, that issue is a bit moot, as it turns out.  You can only make a "concrete frames only" StackFrameList at construction time by passing true to the `show_inlined_frames` argument.  Nowhere in lldb is that passed as false, and I can't see evidence that we ever made that possible.  It is also entirely untested.
I've deleted this code in the patch.
We should probably also discuss whether we need to preserve this "OnlyConcreteFrames" stack feature given that we haven't used it pretty much forever so far as I can see.

labath replied:
Sounds good. Thanks.

That said, I think the code below still needs a read lock as it's accessing m_frames. I think that might be avoidable if `GetFramesUpTo` returned the frame it was asked for (see other suggestions).

jimingham committed:
Move some functions to private where appropriate, and add a few
comments explaining how this is to be used.

jimingham committed:
Reduce the use of m_list_mutex, try to separate reader and writer
lock uses.

jimingham committed:
Take the shared mutex where needed in GetNumFrames.

jimingham commented:
I simplified this as much as I could.  I had started by assuming that the places where we were acquiring m_list_mutex were all required, and the problem was that some were reader and some writer uses but the single mutex type treated them all as writers.  So I divided them by their actual usages and then dealt with the two usages as appropriate.  But in fact in a number of places we were just grabbing the mutex because we might want it and since it was recursive it didn't hurt.

So I took out all the mutex uses that didn't surround code that either directly accesses the `m_frames` or `m_concrete_frames_fetched`.  That reduced some of the noise.  Then since switching roles seemed to concern you, I reworked it so that we only did writer tasks in whole functions now documented to have that role - and had to be called with no lock taken.  The callers that want to do "check whether I have to do work" then "do work" have to acquire the shared lock, do the checking, drop it and call one of the writer functions.

Hopefully this will seem clearer.

labath commented on lldb/include/lldb/Target/StackFrameList.h:
```diff
@@ -93,25 +94,37 @@ class StackFrameList {
-  void SynthesizeTailCallFrames(StackFrame &next_frame);
-
-  bool GetAllFramesFetched() { return m_concrete_frames_fetched == UINT32_MAX; }
+  /// Callers should first check (under the shared mutex) whether we need to
+  /// fetch frames or not.
```
(optional) I'd consider putting the read-locked check into the function itself. I think then you wouldn't need the `WereAllFramesFetched` check in `GetNumFrames`, and I don't think it should hurt the other callers.

jimingham replied:
The two uses of GetFramesUpTo figure out whether they can get what they need w/o adding new frames in different ways.  In general, I like that the immediate callers can figure out by their own lights whether they have the frames they need, and then only call GetFramesUpTo if you need to fetch more frames.  That seems clearer to me.

labath commented on lldb/source/Target/StackFrameList.cpp:
```diff
@@ -578,58 +607,36 @@ void StackFrameList::Dump(Stream *s) {
-    frame_sp = m_frames[idx];
+    if (idx < m_frames.size())
+      frame_sp = m_frames[idx];
 
-  if (frame_sp)
-    return frame_sp;
+    if (frame_sp)
+      return frame_sp;
```
(optional, idea) If `GetFramesUpTo` somehow returned the frame at the given index then there wouldn't be a need for this block here. I'm thinking about this in conjunction with my other suggestion to have a read-locked fastpath in `GetFramesUpTo`.

labath reviewed (commented):
Thank you. Despite the number of comments, I think this looks much better. I've made some suggestions that I think could reduce the number of times the mutex is taken (not so much for the performance, but for the readability benefit). I've marked them as optional, but I think they'd could improve things, and shouldn't be too hard to implement.

jimingham commented on lldb/source/Target/StackFrameList.cpp:
```diff
@@ -803,13 +819,12 @@ uint32_t StackFrameList::SetSelectedFrame(lldb_private::StackFrame *frame) {
-
   SetDefaultFileAndLineToSelectedFrame();
-  return *m_selected_frame_idx;
+  result = *m_selected_frame_idx;
```
That was left over from a previous refactoring.

jimingham commented on lldb/source/Target/StackFrameList.cpp:
```diff
@@ -25,6 +25,7 @@
 #include "lldb/Target/Unwind.h"
 #include "lldb/Utility/LLDBLog.h"
 #include "lldb/Utility/Log.h"
+#include "llvm/ADT/ScopeExit.h"
```
True, that.

jimingham commented on lldb/source/Target/StackFrameList.cpp:
```diff
@@ -119,6 +120,7 @@ bool StackFrameList::DecrementCurrentInlinedDepth() {
     uint32_t current_inlined_depth = GetCurrentInlinedDepth();
     if (current_inlined_depth != UINT32_MAX) {
       if (current_inlined_depth > 0) {
+        std::lock_guard<std::mutex> guard(m_inlined_depth_mutex);
```
That switch was necessitated by the change of `m_list_mutex` from a recursive to a non-recursive mutex.  This code often gets executed under some frame that has the list mutex.  So you can't just leave it as is.

The control of this feature isn't really well modeled by mutexes.  I think they were just there because recursive mutexes make it easy to reflexively put them around anything...

But the inlined depth of the current stack frame has to be set before we present the StackFrameList - it doesn't really make sense to do it otherwise.  BTW, that doesn't violate the "lazy stack frame building" aim because we only compute this for the inlined frames below the first concrete frame, which we always fetch.

So by the time any clients see the StackFrameList, the inlined depth should have been set.  It's not as simple as "has to be done in the constructor" because we also need to reset it at the point where we decide that we can implement a "step in" by just changing the inlined stack counter.  We do that without rebuilding the stack frame.

Currently this isn't a problem because we wrote the code so that it sets the inlined depth in the right place.  So perhaps we should just drop this mutex altogether and rely on this getting set and accessed in the right order, as currently happens?  But I'd rather do that as a separate patch.

labath replied:
Please do that. Someone from the future is going to thank you.

jimingham committed:
Respond to review comments.

jimingham commented:
I pushed the last set of changes implementing some of your suggestions and adding the missing shared lock.

labath commented on lldb/source/Target/StackFrameList.cpp:
```diff
@@ -578,72 +609,53 @@ void StackFrameList::Dump(Stream *s) {
-              frame_sp->GetSymbolContext(eSymbolContextSymbol).symbol);
-        }
-        SetFrameAtIndex(idx, frame_sp);
+  {  // Now we're accessing m_frames as a reader, so acquire the reader lock.
```
This scope shouldn't be necessary. If I'm reading this correctly, we're returning directly afterwards.

labath reviewed (approved):
Thanks for your patience.

jimingham committed:
Last round of clang-format.

mstorsjo commented:
FYI, this had a small impact on deployability on macOS; I used to build LLVM with 10.9 as a deployment target, but now get these errors:

```
/Users/runner/work/llvm-mingw/llvm-mingw/llvm-project/lldb/include/lldb/Target/StackFrameList.h:161:16: error: 'shared_mutex' is unavailable: introduced in macOS 10.12
  mutable std::shared_mutex m_list_mutex;
               ^
```

Raising the deployment target to 10.12 instead of 10.9 is not an issue for me, it's more than plenty old for where anybody might want to run my toolchain. But IIRC there are some downstream distributions that do build recent LLVM for very old versions of macOS too (but perhaps they provide a newer libc++ to go with it, I don't know).

So - not an issue (yet AFAIK), but just a note that this does change the potential minimum baseline.

jimingham commented:
When we decided to allow C++11 usage in lldb I think we opted into requiring platforms that have relatively complete support for the C++11 additions.  I don't think it makes sense to try to limit the use of C++11 to some "commonly supported subset".

mstorsjo commented:
> When we decided to allow C++11 usage in lldb I think we opted into requiring platforms that have relatively complete support for the C++11 additions. I don't think it makes sense to try to limit the use of C++11 to some "minimal supported subset".

Yeah, I don’t disagree with that - and we’re at C++17 now already. Anyway, requiring 10.12 from 2016 seems reasonable to me.

adrian-prantl commented:
> Yeah, I don’t disagree with that - and we’re at C++17 now already. Anyway, requiring 10.12 from 2016 seems reasonable to me.

Especially since nobody needs to run LLDB locally in order to do debugging. If you really want to debug with the very latest LLDB on a 10 year old system, you can debug remotely.
